src/validator/parent_visitor.rs

```diff
@@ -0,0 +1,1030 @@
+    if let Some(CDDLType::GenericParams(params)) = CDDLType::from(self).parent(parent_visitor) {
+      return Some(params);
+    }
+
+    None
+  }
+}
+
+impl<'a, 'b: 'a> Parent<'a, 'b, Type2<'a>> for GenericArgs<'a> {
+  fn parent(&'a self, parent_visitor: &'b ParentVisitor<'a, 'b>) -> Option<&Type2<'a>> {
+    if let Some(CDDLType::Type2(t2)) = CDDLType::from(self).parent(parent_visitor) {
+      return Some(t2);
+    }
+
+    None
+  }
+}
+
+impl<'a, 'b: 'a> Parent<'a, 'b, GenericArgs<'a>> for GenericArg<'a> {
+  fn parent(&'a self, parent_visitor: &'b ParentVisitor<'a, 'b>) -> Option<&GenericArgs<'a>> {
+    if let Some(CDDLType::GenericArgs(args)) = CDDLType::from(self).parent(parent_visitor) {
+      return Some(args);
+    }
+
+    None
+  }
+}
+
+impl<'a, 'b: 'a> Parent<'a, 'b, Type1<'a>> for Operator<'a> {
+  fn parent(&'a self, parent_visitor: &'b ParentVisitor<'a, 'b>) -> Option<&Type1<'a>> {
+    if let Some(CDDLType::Type1(t1)) = CDDLType::from(self).parent(parent_visitor) {
+      return Some(t1);
+    }
+
+    None
+  }
+}
+
+impl<'a, 'b: 'a> Parent<'a, 'b, Type2<'a>> for Type<'a> {
+  fn parent(&'a self, parent_visitor: &'b ParentVisitor<'a, 'b>) -> Option<&Type2<'a>> {
+    if let Some(CDDLType::Type2(t2)) = CDDLType::from(self).parent(parent_visitor) {
+      return Some(t2);
+    }
+
+    None
+  }
+}
+
+#[derive(Debug, Default, Clone)]
+struct ArenaTree<'a, 'b: 'a> {
+  arena: Vec<Node<'a, 'b>>,
+}
+
+impl<'a, 'b: 'a> ArenaTree<'a, 'b> {
+  fn node(&mut self, val: CDDLType<'a, 'b>) -> usize {
+    for node in self.arena.iter() {
+      if node.val == val {
+        return node.idx;
+      }
+    }
+
+    let idx = self.arena.len();
+    self.arena.push(Node::new(idx, val));
+    idx
+  }
+}
+
+#[derive(Debug, Clone)]
+struct Node<'a, 'b: 'a> {
+  idx: usize,
+  val: CDDLType<'a, 'b>,
+  parent: Option<usize>,
+  children: Vec<usize>,
+}
+
+impl<'a, 'b: 'a> Node<'a, 'b> {
+  fn new(idx: usize, val: CDDLType<'a, 'b>) -> Self {
+    Self {
+      idx,
+      val,
+      parent: None,
+      children: vec![],
+    }
+  }
+}
+
+/// validator type
+// #[derive(Clone)]
+pub struct ParentVisitor<'a, 'b: 'a> {
+  arena_tree: ArenaTree<'a, 'b>,
+}
+
+impl<'a, 'b: 'a> ParentVisitor<'a, 'b> {
+  pub fn new(cddl: &'a CDDL<'a>) -> Result<Self> {
+    let mut p = ParentVisitor {
+      arena_tree: ArenaTree {
+        arena: Vec::default(),
+      },
+    };
+
+    p.visit_cddl(cddl)?;
+
+    Ok(p)
+  }
+}
+
+impl<'a, 'b: 'a> ParentVisitor<'a, 'b> {
+  fn insert(&mut self, parent: usize, child: usize) -> Result<()> {
+    match self.arena_tree.arena[child].parent {
+      Some(_) => {
+        // return Err(Error::Overwrite);
+      }
+      None => {
+        self.arena_tree.arena[child].parent = Some(parent);
+      }
+    }
+
+    self.arena_tree.arena[parent].children.push(child);
+
+    Ok(())
+  }
+}
+
+impl<'a, 'b: 'a> CDDLType<'a, 'b> {
+  fn parent(&self, visitor: &'b ParentVisitor<'a, 'b>) -> Option<&'b CDDLType<'a, 'b>> {
```

Comment:
I assume this method is meant to be public?

Reply:
How do I get the parent of a CDDLType representing an unknown AST element though?

Reply:
Good point. Will make it public.